tests/pthread/test_pthread_proxying_cpp.cpp

```diff
@@ -0,0 +1,158 @@
+std::atomic<bool> should_quit{false};
+
+// Whether `returner` has spun up.
+std::atomic<bool> has_begun{false};
+
+void looper_main() {
+  while (!should_quit) {
+    queue.execute();
+    sched_yield();
+  }
+}
+
+void returner_main() {
+  has_begun = 1;
+  emscripten_exit_with_live_runtime();
+}
+
+void test_proxy_async() {
+  std::cout << "Testing async proxying\n";
+
+  int i = 0;
+
+  std::mutex mutex;
+  std::condition_variable cond;
+  std::thread::id executor;
+
+  // Proxy to ourselves.
+  queue.proxyAsync(pthread_self(), [&]() {
+    i = 1;
+    executor = std::this_thread::get_id();
+  });
+  assert(i == 0);
+  queue.execute();
+  assert(i == 1);
+  assert(executor == std::this_thread::get_id());
+
+  // Proxy to looper.
+  {
+    queue.proxyAsync(looper.native_handle(), [&]() {
+      i = 2;
+      executor = std::this_thread::get_id();
+      cond.notify_one();
+    });
+    std::unique_lock<std::mutex> lock(mutex);
+    cond.wait(lock, [&]() { return i == 2; });
+    assert(executor == looper.get_id());
+  }
+
+  // Proxy to returner.
+  {
+    queue.proxyAsync(returner.native_handle(), [&]() {
+      i = 3;
+      executor = std::this_thread::get_id();
+      cond.notify_one();
+    });
+    std::unique_lock<std::mutex> lock(mutex);
+    cond.wait(lock, [&]() { return i == 3; });
+    assert(executor == returner.get_id());
+  }
+}
+
+void test_proxy_sync() {
+  std::cout << "Testing sync proxying\n";
+
+  int i = 0;
+  std::thread::id executor;
+
+  // Proxy to looper.
+  {
+    queue.proxySync(looper.native_handle(), [&]() {
+      i = 2;
+      executor = std::this_thread::get_id();
+    });
+    assert(i == 2);
+    assert(executor == looper.get_id());
+  }
+
+  // Proxy to returner.
+  {
+    queue.proxySync(returner.native_handle(), [&]() {
+      i = 3;
+      executor = std::this_thread::get_id();
+    });
+    assert(i == 3);
+    assert(executor == returner.get_id());
+  }
+}
+
+void test_proxy_sync_with_ctx(void) {
+  std::cout << "Testing sync proxying\n";
+
+  int i = 0;
+  std::thread::id executor;
+
+  // Proxy to looper.
+  {
+    queue.proxySyncWithCtx(looper.native_handle(), [&](auto ctx) {
+      i = 2;
+      executor = std::this_thread::get_id();
+      ctx.finish();
+    });
+    assert(i == 2);
+    assert(executor == looper.get_id());
+  }
+
+  // Proxy to returner.
+  {
+    queue.proxySyncWithCtx(returner.native_handle(), [&](auto ctx) {
+      i = 3;
+      executor = std::this_thread::get_id();
+      ctx.finish();
+    });
+    assert(i == 3);
+    assert(executor == returner.get_id());
+  }
+}
+
+int main(int argc, char* argv[]) {
+  looper = std::thread(looper_main);
+  returner = std::thread(returner_main);
+
+  // `returner` can't process its queue until it starts up.
+  while (!has_begun) {
+    sched_yield();
+  }
+
+  test_proxy_async();
+  test_proxy_sync();
+  test_proxy_sync_with_ctx();
+
+  should_quit = true;
+  looper.join();
+
+  pthread_cancel(returner.native_handle());
```

Comment:
Seems odd to use mix the pthread and C++ thread API like this.   Cam you not post a message to the thread and have it exit on its own... Oh.. maybe we are missing the API that lets a thread force exit just the current thread?  (i.e. override exit_with_live_runtime?)  Is so that seems like a missing API that we should add.. maybe add a TODO here with a link to a bug?

Reply:
The problem here is that there's no portable way to cancel a `std::thread` O.o Posting a message to the thread to make it cancel would work, but seems like overkill for the test.

Reply:
I don't think we want to be cancelling threads.. even in test code. and especially when it involves looking though the C++ thread abstraction.  I think there probably a reason that cancellation was not included in the C++11 thread API.

How about just exiting without joining and letting the OS (emscripten) tear down the thread?

Reply:
The threads just hang indefinitely without the explicit cancel :disappointed:  (Also FWIW C++20 adds thread cancellation with a new `std::jthread`.)

Reply:
We support C++20.  Do you want to try in this test?  :)   (only half joking).

Reply:
We should also fix the indefinite hang.. can you open a bug for that?

Reply:
Looks like we don't support `std::jthread` yet :(

Issue about the hang filed: #16493